File: py-server/src/chat/stream_emitter.py
import asyncio
import json
from typing import Any, AsyncIterator


class StreamEmitter:
    """SSE 发射器：queue 桥接同步 send() 与异步 HTTP 流。
    与 Nest 的 res.write 对齐，send() 是同步调用。
    """

    def __init__(self):
        self._queue: asyncio.Queue[str | None] = asyncio.Queue()

    def send(self, event: dict[str, Any]) -> None:
        # default=str 与 Nest 的 JSON.stringify 对齐：datetime/date/Decimal 等
        # DB 常返类型统一转字符串，否则 json.dumps 会抛 "not JSON serializable"
        self._queue.put_nowait(
            f"data: {json.dumps(event, ensure_ascii=False, default=str)}\n\n"
        )

    def done(self) -> None:
        self.send({"type": "done"})
        self._queue.put_nowait(None)

    def end(self) -> None:
        """终止流（不发 done 事件，用于强制关闭场景）。"""
        self._queue.put_nowait(None)

    async def stream(self) -> AsyncIterator[str]:
        while True:
            chunk = await self._queue.get()
            if chunk is None:
                break
            yield chunk

File: java-server/src/main/java/com/gistagent/chat/StreamEmitter.java
package com.gistagent.chat;

import java.io.IOException;
import java.util.LinkedHashMap;
import java.util.Map;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter;

/**
 * SSE 推送器：封装 Spring {@link SseEmitter}，对齐 Nest 版 StreamEmitter 的线协议
 * ({@code data: <json>\n\n})。
 *
 * 设计取舍：
 * - Nest 端通过裸写 {@code res.write('data: ' + JSON + '\n\n')} 手动构造 SSE，
 *   Spring 默认的 {@link SseEmitter#event()} 会加 {@code id:}/{@code event:} 行，
 *   前端已有解析逻辑只认 {@code data:} 一行，所以这里用 {@code send(Object)} 直发。
 * - {@link #send(SseEvent)} 捕获 IOException 只记日志——客户端断开是常态，
 *   不应当把异常冒泡到 Agent Loop 导致对话整体失败。
 */
public final class StreamEmitter {

	private static final Logger log = LoggerFactory.getLogger(StreamEmitter.class);

	private final SseEmitter emitter;
	private final ObjectMapper mapper;

	public StreamEmitter(SseEmitter emitter, ObjectMapper mapper) {
		this.emitter = emitter;
		this.mapper = mapper;
	}

	public void send(SseEvent event) {
		try {
			emitter.send(event.payload());
		} catch (IOException e) {
			log.debug("SSE send failed (client closed?): {}", e.getMessage());
		}
	}

	public void done() {
		send(SseEvent.of("done"));
		emitter.complete();
	}

	public void fail(Throwable t) {
		try {
			Map<String, Object> err = new LinkedHashMap<>();
			err.put("type", "error");
			err.put("content", t.getMessage() == null ? "unknown error" : t.getMessage());
			emitter.send(err);
		} catch (IOException e) {
			log.debug("SSE error event failed: {}", e.getMessage());
		}
		emitter.complete();
	}

	/**
	 * 用于构造 {@link SseEvent}：提供“安全字符串化”入口，避免调用方自行
	 * {@code ObjectMapper} 拼接出错。
	 */
	public String stringify(Object value) {
		try {
			return mapper.writeValueAsString(value);
		} catch (JsonProcessingException e) {
			return "\"\"";
		}
	}
}
